go/registry/registry.go

```diff
@@ -0,0 +1,239 @@
+	return func(config *Config) {
+		config.apiBaseUrl = apiBaseUrl
+	}
+}
+
+func WithOrganization(organization string) Option {
+	return func(config *Config) {
+		config.organization = organization
+	}
+}
+
+// Create a new runtime for a specific scalefile
+func New(function string, tag string, opts ...Option) (*scalefunc.ScaleFunc, error) {
+	// Default config
+	conf := &Config{
+		pullPolicy:     AlwaysPullPolicy,
+		cacheDirectory: "~/.cache/scale/functions",
+		apiKey:         "",
+		apiBaseUrl:     "https://api.scale.sh/v1",
+		organization:   "default",
+	}
+	for _, opt := range opts {
+		opt(conf)
+	}
+
+	// First check our local cache...
+	sf, err := getFromCache(function, tag, conf)
+
+	if err == nil && conf.pullPolicy != AlwaysPullPolicy {
+		return sf, err
+	}
+
+	if conf.pullPolicy == NeverPullPolicy {
+		return nil, ErrNoFunction
+	}
+
+	// Contact the API endpoint with the request
+	response, err := apiRequest(function, tag, conf)
+	if err != nil {
+		return sf, err
+	}
+
+	// Get the scalefunc from the URL
+	httpResp, err := http.Get(response.PresignedURL)
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := io.ReadAll(httpResp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	h := sha256.New()
+	h.Write(data)
+
+	bs := h.Sum(nil)
+
+	s := base64.URLEncoding.EncodeToString(bs)
+
+	if s != response.Hash {
+		return nil, ErrHashMismatch
+	}
+
+	// Save to our local cache
+	err = saveToCache(function, tag, conf, data)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode to a scalefile
+	err = sf.Decode(data)
+	if err != nil {
+		return nil, err
+	}
+	return sf, nil
+}
+
+// build a filename from the config
+// TODO: We should use the hash in the filename for optimization
+func buildFilename(function string, tag string, conf *Config) string {
+	return fmt.Sprintf("%s.%s.scale", function, tag)
+}
+
+// Get a scalefile from the local cache
+func getFromCache(function string, tag string, conf *Config) (*scalefunc.ScaleFunc, error) {
+	f := buildFilename(function, tag, conf)
+
+	// Try to read the scalefile
+	path := fmt.Sprintf("%s%c%s", conf.cacheDirectory, os.PathSeparator, f)
+	return scalefunc.Read(path)
+}
+
+// Save a scalefile to our local cache
+func saveToCache(function string, tag string, conf *Config, data []byte) error {
+	err := os.MkdirAll(conf.cacheDirectory, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Overwrite the file
+	f := buildFilename(function, tag, conf)
+	path := fmt.Sprintf("%s%c%s", conf.cacheDirectory, os.PathSeparator, f)
+
+	fh, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0755)
+	if err != nil {
+		return err
+	}
+
+	_, err = fh.Write(data)
+	if err != nil {
+		return err
+	}
+
+	err = fh.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func removeCache(conf *Config) error {
+	return os.RemoveAll(conf.cacheDirectory)
+}
+
+// PulldownResponse API response when requesting a function
+type GetFunctionResponse struct {
```

Comment:
We should probably use a swagger-generated client if we can here, but this is fine for now